sdcm/scan_operation_thread.py

```diff
@@ -1,163 +1,316 @@
+    total_scan_time: int = 0
+    stats: list[FullScanOperationStat] = field(default_factory=list)
+
+    def get_stats_pretty_table(self) -> PrettyTable | None:
+        if not self.stats:
+            return None
+
+        pretty_table = PrettyTable(field_names=[field.name for field in fields(self.stats[0])])
+        for stat in self.stats:
+            pretty_table.add_row([stat.op_type, stat.duration, stat.exceptions, stat.nemesis_at_start,
+                                  stat.nemesis_at_end, stat.success, stat.cmd])
+        return pretty_table
+
+
+@dataclass
+class FullScanOperationStat:
+    """
+    Keeps track of stats for a single fullscan operation.
+    """
+    op_type: str = None
+    duration: float = None
+    exceptions: list = field(default_factory=list)
+    nemesis_at_start: str = None
+    nemesis_at_end: str = None
+    success: bool = None
+    cmd: str = None
+
+
+class FullScanCommand(NamedTuple):
+    name: str
+    base_query: Template
+
+
+class FullScanAggregateCommands(NamedTuple):
+    SELECT_ALL = FullScanCommand("SELECT_ALL", Template("SELECT * from $ks_cf$bypass_cache$timeout"))
+    AGG_COUNT_ALL = FullScanCommand("AGG_COUNT_ALL", Template("SELECT count(*) FROM $ks_cf$timeout$bypass_cache"))
+
+
+class FullscanException(Exception):
+    """ Exception during running a fullscan"""
+    ...
+
+
 # pylint: disable=too-many-instance-attributes
 class ScanOperationThread:
-    bypass_cache = ' bypass cache'
-    basic_query = 'select * from {}'
-
-    # pylint: disable=too-many-arguments,unused-argument
-    def __init__(self, db_cluster: [BaseScyllaCluster, BaseCluster], duration: int, interval: int,
-                 termination_event: threading.Event, ks_cf: str,
-                 scan_event: [FullScanEvent, FullPartitionScanReversedOrderEvent], page_size: int = 10000, **kwargs):
-        self.ks_cf = ks_cf
-        self.db_cluster = db_cluster
-        self.page_size = page_size
-        self.duration = duration
-        self.interval = interval
-        self.number_of_rows_read = 0
-        self.db_node = None
-        self.read_pages = 0
-        self.scans_counter = 0
-        self.time_elapsed = 0
-        self.total_scan_time = 0
-        self.scan_event = scan_event
-        self.termination_event = termination_event
+    """
+    Runs fullscan operations according to the parameters specified in the test
+    config yaml files. Has 4 main modes:
+    - random: uses a seeded random generator to generate a queue of fullscan
+    operations using all the available operation types
+    - select: uses only FullScanOperation
+    - partition: uses only FullPartitionScanOperation
+    - aggregate: uses only FullScanAggregatesOperation
+
+    Check FullScanParams class for parameters to tweak.
+
+    Note: the seeded random generator is shared between the thread class and
+    the operations instances, so the generated CQL commands are also replicable
+    when the same seed value is used.
```

Comment:
The operations are instances of the *Operation classes in this module. They are added randomly to a queue at init time, and then popped from the queue one by one when the thread is running. Since each instance creates it's own CQL query with some randomized elements, I wanted them to also do it in a reproducible way (no use of having a reproducible sequence of fullscans if the commands differ from run to run).
I updated the docstring to hopefully make it more clear.